File: arrays/8. next permutation.cpp
class Solution {
public:
    void reverse(vector<int>& nums, int lo, int hi){
        while(lo<=hi){
            swap(nums[lo], nums[hi]); 
            lo++; hi--;
        }
    }
    
    void nextPermutation(vector<int>& nums) {
        int n=nums.size();
        int i;
        for(i=n-2;i>=0;i--){
            if (nums[i] < nums[i+1]){
                break;
            }
        }
        if (i==-1){
            reverse(nums,0,n-1);
        }
        else{
            for(int k=n-1;k>i;k--){
                if (nums[k] > nums[i]){
                    swap(nums[k],nums[i]);
                    reverse(nums,i+1,n-1);
                    return;
                }
            }
        }
    }
};

File: Python/backtracking/8. next_permutation.py
# https://leetcode.com/problems/next-permutation/description/


class Solution:
    def reverse(self, lo, hi, nums):
        while lo<=hi:
            nums[lo], nums[hi] = nums[hi], nums[lo]
            lo+=1
            hi-=1
        
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # find the inflection point
        n, idx = len(nums), -1

        for i in range(n-1, 0, -1):
            if nums[i-1] < nums[i]:
                idx = i-1
                break
        
        if idx==-1:
            self.reverse(0,n-1,nums)
            return
        
        print(idx)
        # find the first element greater that nums[idx]
        for j in range(n-1,idx,-1):
            if nums[j] > nums[idx]:
                # swap nums[i], nums[j]
                nums[i], nums[j] = nums[j], nums[i]
                self.reverse(idx+1,n-1,nums)
                return

# TC: O(n)
# SC: O(1)